{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import needed libraries.\n",
    "\n",
    "- BeautifulSoup is used for web scrapping purposes\n",
    "\n",
    "- requests, urllib, and ssl are used for making api request\n",
    "\n",
    "- json used to deal with json format\n",
    "\n",
    "- pandas used for dealing with data and exporting\n",
    "\n",
    "Potential Additions: Extracting Keywords and Citations\n",
    "\n",
    "- However citations just give another article ID."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from bs4 import BeautifulSoup\n",
    "import requests, urllib, ssl, json\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "API Request Key"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "api_req_url=\"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pubmed&retmode=json&retmax=100&sort=relevance&term=\"\n",
    "keyword = \"KEYWORD\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ask for user input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter the keyword required: \n",
      "immport\n"
     ]
    }
   ],
   "source": [
    "keyword = str(input(\"Enter the keyword required: \\n\"))\n",
    "api_req_url = api_req_url + keyword"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Need to verify your HTTP Certificate and then make request."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "_create_unverified_https_context = ssl._create_unverified_context\n",
    "request = urllib.request.urlopen(api_req_url).read()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now need to be able to access JSON data and parse specific things\n",
    "\n",
    "TODO: Could add a feature for inputting how many articles wanted and also wanted/totalmatch ratio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['29485622',\n",
       " '32411773',\n",
       " '24791905',\n",
       " '31272390',\n",
       " '32504750',\n",
       " '28057685',\n",
       " '30852461',\n",
       " '30977847',\n",
       " '30304689',\n",
       " '32600423',\n",
       " '32375704',\n",
       " '32351547',\n",
       " '32310824',\n",
       " '32068352',\n",
       " '32355728',\n",
       " '32296685',\n",
       " '32214002',\n",
       " '32019546',\n",
       " '31921619',\n",
       " '31821170',\n",
       " '30832680',\n",
       " '30594555',\n",
       " '30516029',\n",
       " '30398470',\n",
       " '27377652',\n",
       " '27130330',\n",
       " '20839340',\n",
       " '26112029',\n",
       " '20460173',\n",
       " '32283555',\n",
       " '28851925',\n",
       " '25857935',\n",
       " '24862987',\n",
       " '24161793',\n",
       " '22959960',\n",
       " '20044059']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = json.loads(request)\n",
    "\n",
    "article_ids = data['esearchresult']['idlist']\n",
    "article_ids"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Main logic here.\n",
    "\n",
    "Basically scrapes the API return and then gets key information including the abstract that can be used later"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_elements(soup_object):\n",
    "    \n",
    "    authors = getAuthors(soup_object)\n",
    "    title = getTitle(soup_object)\n",
    "    abstract = getAbstract(soup_object)\n",
    "    pubmed_id = getID(soup_object)\n",
    "    \n",
    "    journal_volume = \"\"\n",
    "    journal_title = \"\"\n",
    "    journal_date = \"\"\n",
    "    \n",
    "    if soup_object.find(\"JournalIssue\"):\n",
    "        if soup_object.find(\"JournalIssue\").find(\"Volume\"):\n",
    "            journal_volume = soup_object.find(\"JournalIssue\").find(\"Volume\").text\n",
    "    if soup_object.find(\"Title\"):\n",
    "        journal_title = soup_object.find(\"Title\").text\n",
    "    if soup_object.find(\"JournalIssue\").find(\"Year\") and soup_object.find(\"JournalIssue\").find(\"Month\") and soup_object.find(\"JournalIssue\").find(\"Day\"):\n",
    "        journal_date = soup_object.find(\"JournalIssue\").find(\"Year\").text + \" \" + soup_object.find(\"JournalIssue\").find(\"Month\").text + \" \" + soup_object.find(\"JournalIssue\").find(\"Day\").text\n",
    "    \n",
    "    row_list = []\n",
    "    row_list.append(authors)\n",
    "    row_list.append(title)\n",
    "    row_list.append(journal_title)\n",
    "    row_list.append(journal_volume)\n",
    "    row_list.append(journal_date)\n",
    "    row_list.append(pubmed_id)\n",
    "    row_list.append(abstract)\n",
    "    \n",
    "    return row_list\n",
    "    \n",
    "def getAuthors(soup_object):\n",
    "    authors = \"\"\n",
    "    authorlist = soup_object.find(\"AuthorList\")\n",
    "    ##print(authorlist)\n",
    "    \n",
    "    if authorlist: \n",
    "        for i in range(len(authorlist.find_all(\"LastName\"))):\n",
    "            firstName = authorlist.find_all(\"ForeName\")[i].text\n",
    "            authors = authors + firstName\n",
    "            lastName = authorlist.find_all(\"LastName\")[i].text\n",
    "            authors = authors + \" \" + lastName\n",
    "            \n",
    "            if i != len(authorlist.find_all(\"LastName\")) - 1:\n",
    "                authors = authors + \", \"\n",
    "    print(authors)\n",
    "    return authors\n",
    "\n",
    "def getTitle(soup_object):\n",
    "    title = \"\"\n",
    "    if soup_object.find(\"AricleTitle\"):\n",
    "        title_element = soup_object.find(\"ArticleTitle\")\n",
    "        title = title_element.text\n",
    "\n",
    "    print(title)\n",
    "    return title\n",
    "    \n",
    "def getAbstract(soup_object):\n",
    "    abstract = \"\"\n",
    "    if soup_object.find(\"AbstractText\"):\n",
    "        abstract = soup_object.find(\"AbstractText\").text\n",
    "    print(abstract)\n",
    "    return abstract\n",
    "\n",
    "def getID(soup_object):\n",
    "    idInfo = \"\"\n",
    "    if soup_object.find(\"PMID\"):\n",
    "        idInfo = \"PMID: \"\n",
    "        idInfo += soup_object.find(\"PMID\").text\n",
    "    return idInfo"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Loops through all the different IDs and the scraping material that was gathered using the above logic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sanchita Bhattacharya, Patrick Dunn, Cristel G Thomas, Barry Smith, Henry Schaefer, Jieming Chen, Zicheng Hu, Kelly A Zalocusky, Ravi D Shankar, Shai S Shen-Orr, Elizabeth Thomson, Jeffrey Wiser, Atul J Butte\n",
      "\n",
      "Immunology researchers are beginning to explore the possibilities of reproducibility, reuse and secondary analyses of immunology data. Open-access datasets are being applied in the validation of the methods used in the original studies, leveraging studies for meta-analysis, or generating new hypotheses. To promote these goals, the ImmPort data repository was created for the broader research community to explore the wide spectrum of clinical and basic research data and associated findings. The ImmPort ecosystem consists of four components-Private Data, Shared Data, Data Analysis, and Resources-for data archiving, dissemination, analyses, and reuse. To date, more than 300 studies have been made freely available through the Shared Data portal (www.immport.org/immport-open), which allows research data to be repurposed to accelerate the translation of new insights into discoveries.\n",
      "Lei Sun, Zhe Zhang, Yao Yao, Wen-Ya Li, Jia Gu\n",
      "\n",
      "There has been little investigation carried out into the activity of immune-related genes in the prognosis of non-small cell lung cancer (NSCLC). Our study set out to analyze the correlation between the differential expression of immune genes and NSCLC prognosis by screening the differential expression of immune genes. Based on the immune genes identified, we aimed to construct a prognostic risk model and explore some novel molecules which have predictive potential for therapeutic effect and prognosis in lung cancer.\n",
      "Sanchita Bhattacharya, Sandra Andorf, Linda Gomes, Patrick Dunn, Henry Schaefer, Joan Pontius, Patty Berger, Vince Desborough, Tom Smith, John Campbell, Elizabeth Thomson, Ruth Monteiro, Patricia Guimaraes, Bryan Walters, Jeff Wiser, Atul J Butte\n",
      "\n",
      "The immunology database and analysis portal (ImmPort) system is the archival repository and dissemination vehicle for clinical and molecular datasets created by research consortia funded by the National Institute of Allergy and Infectious Diseases Division of Allergy, Immunology, and Transplantation. With nearly 100 datasets now publicly available and hundreds of downloads per month, ImmPort is an important source for raw data and protocols from clinical trials, mechanistic studies, and novel methods for cellular and molecular measurements. To facilitate data transfer, templates for data representation and standard operating procedures have also been created and are also publicly available. ImmPort facilitates transparency and reproducibility in immunology research, serves as an important resource for education, and enables newly generated hypotheses and data-driven science.\n",
      "James A Overton, Randi Vita, Patrick Dunn, Julie G Burel, Syed Ahmad Chan Bukhari, Kei-Hoi Cheung, Steven H Kleinstein, Alexander D Diehl, Bjoern Peters\n",
      "\n",
      "Human immunology studies often rely on the isolation and quantification of cell populations from an input sample based on flow cytometry and related techniques. Such techniques classify cells into populations based on the detection of a pattern of markers. The description of the cell populations targeted in such experiments typically have two complementary components: the description of the cell type targeted (e.g. 'T cells'), and the description of the marker pattern utilized (e.g. CD14-, CD3+).\n",
      "Jingrun Yang, Kuixia Xie, Chengxin Li\n",
      "\n",
      "Head and neck squamous cell carcinoma (HNSCC) is an highly aggressive tumor with heterogeneous prognosis. We here report that immune-related genes (IRGs) could effectively distinguish prognostically different HNSCC patients.\n",
      "Ravi D Shankar, Sanchita Bhattacharya, Chethan Jujjavarapu, Sandra Andorf, Jeffery A Wiser, Atul J Butte\n",
      "\n",
      ": Open access to raw clinical and molecular data related to immunological studies has created a tremendous opportunity for data-driven science. We have developed RImmPort that prepares NIAID-funded research study datasets in ImmPort (immport.org) for analysis in R. RImmPort comprises of three main components: (i) a specification of R classes that encapsulate study data, (ii) foundational methods to load data of a specific study and (iii) generic methods to slice and dice data across different dimensions in one or more studies. Furthermore, RImmPort supports open formalisms, such as CDISC standards on the open source bioinformatics platform Bioconductor, to ensure that ImmPort curated study datasets are seamlessly accessible and ready for analysis, thus enabling innovative bioinformatics research in immunology.\n",
      "Sandra Andorf, Sanchita Bhattacharya, Brice Gaudilliere, Gary M Shaw, David K Stevenson, Atul J Butte, Marina Sirota\n",
      "\n",
      "Preterm birth (PTB), or the delivery of an infant prior to 37 weeks of gestation, is a major health concern. Although a variety of social, environmental, and maternal factors have been implicated in PTB, causes of preterm labor have remained largely unknown. There is evidence of effectiveness and safety of influenza vaccination during pregnancy, however fewer studies have looked at vaccination response as an indicator of an innate host response that may be associated with adverse pregnancy outcomes. We carried out a pilot study to analyze the flu vaccine response during pregnancy of women who later deliver preterm or term.\n",
      "Jieming Chen, Sanchita Bhattacharya, Marina Sirota, Sunisa Laiudompitak, Henry Schaefer, Elizabeth Thomson, Jeff Wiser, Minnie M Sarwal, Atul J Butte\n",
      "\n",
      "There are limited resources providing postdonation conditions that can occur in living donors (LDs) of solid-organ transplant. Consequently, it is difficult to visualize and understand possible postdonation outcomes in LDs.\n",
      "Kelly A Zalocusky, Matthew J Kan, Zicheng Hu, Patrick Dunn, Elizabeth Thomson, Jeffrey Wiser, Sanchita Bhattacharya, Atul J Butte\n",
      "\n",
      "There is increasing appreciation that the immune system plays critical roles not only in the traditional domains of infection and inflammation but also in many areas of biology, including tumorigenesis, metabolism, and even neurobiology. However, one of the major barriers for understanding human immunological mechanisms is that immune assays have not been reproducibly characterized for a sufficiently large and diverse healthy human cohort. Here, we present the 10,000 Immunomes Project (10KIP), a framework for growing a diverse human immunology reference, from ImmPort, a publicly available resource of subject-level immunology data. Although some measurement types are sparse in the presently deposited ImmPort database, the extant data allow for a diversity of robust comparisons. Using 10KIP, we describe variations in serum cytokines and leukocytes by age, race, and sex; define a baseline cell-cytokine network; and describe immunologic changes in pregnancy. All data in the resource are available for visualization and download at http://10kimmunomes.org/.\n",
      "Sihan Chen, G D Cao, Wu Wei, Lu Yida, He Xiaobo, Yang Lei, Chen Ke, Bo Chen, Mao Ming Xiong\n",
      "\n",
      "Colon adenocarcinoma (COAD) is a gastrointestinal tumor with a high degree of malignancy. Its deterioration process is closely related to the tumor microenvironment, and transcription factors (TF) play a regulatory role in this process. Currently, there is a lack of exploration between the genes related to the COAD tumor microenvironment and the survival prognosis of patients. Models composed of multiple genes usually predict the survival prognosis of patients more accurately than single genes. We can analyze the multigene models that can predict the prognosis of COAD from the current database.\n",
      "Haitao Chen, Jun Luo, Jianchun Guo\n",
      "\n",
      "Colon cancer is a common and highly malignant cancer. Its morbidity is rapidly increasing, and its prognosis is poor. Currently, immunotherapy is a rapidly developing therapeutic modality of colon cancer. This study aimed to construct a prognostic risk model based on immune genes for the early diagnosis and accurate prognostic prediction of colon cancer.\n",
      "Mingwei Zhang, Xuezhen Wang, Xiaoping Chen, Qiuyu Zhang, Jinsheng Hong\n",
      "\n",
      "Despite several clinicopathological factors being integrated as prognostic biomarkers, the individual variants and risk stratification have not been fully elucidated in lower grade glioma (LGG). With the prevalence of gene expression profiling in LGG, and based on the critical role of the immune microenvironment, the aim of our study was to develop an immune-related signature for risk stratification and prognosis prediction in LGG.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Rongzhi Huang, Min Mao, Yunxin Lu, Qingliang Yu, Liang Liao\n",
      "\n",
      "Melanoma is a cancer of the skin with potential to spread to other organs and is responsible for most deaths due to skin cancer. It is imperative to identify immune biomarkers for early melanoma diagnosis and treatment.\n",
      "Xiao-Yan Sun, Shi-Zhe Yu, Hua-Peng Zhang, Jie Li, Wen-Zhi Guo, Shui-Jun Zhang\n",
      "\n",
      "Hepatocellular carcinoma (HCC) has become the second most common tumor type that contributes to cancer-related death worldwide. The study aimed to establish a robust immune-related gene pair (IRGP) signature for predicting the prognosis of HCC patients.\n",
      "Weijie Qiang, Yifei Dai, Guibo Sun, Xiaoyan Xing, Xiaobo Sun\n",
      "\n",
      "Colon adenocarcinoma (COAD) is one of the most commonly diagnosed cancers, and it is closely related to the immune microenvironment. Considering that immunotherapy is not effective for all COAD patients, it is necessary to identify the effective population before administering treatment. In this study, we established an independent prognostic index based on immune-related genes (IRGs), in order to evaluate the clinical outcome of COAD.\n",
      "Yang Sheng, Cheng Yanping, Liu Tong, Liu Ning, Liu Yufeng, Liang Geyu\n",
      "\n",
      "Melanoma is a highly aggressive cancer, attracting increasing attention worldwide. The 5-year survival rate of patients with metastatic melanoma is low. Therefore, it is critical to identify potential effective biomarkers for diagnosis of melanoma metastasis. In the present study, the melanoma cohort and immune genes were obtained from the Cancer Genome Atlas (TCGA) database and the ImmPort database, respectively. Then, we constructed the immune risk score (IRS) using univariate and multivariate logistic analysis. The area under the curve (AUC) of IRS in sequencing samples and the initial diagnosis patients was 0.90 and 0.80, respectively. Besides, IRS could add benefits for metastasis diagnosis. For sequencing samples, IRS (OR = 16.35, 95% CI = 8.74-30.59) increased the odds for melanoma metastasis. Similar results were obtained in the initial diagnosis patients (OR = 8.93, 95% CI = 3.53-22.61). A composite nomogram was built based on IRS and clinical information with well-fitted calibration curves. We further used other independent melanoma cohorts from Gene Expression Omnibus (GEO) databases to confirm the reliability and validity of the IRS (AUC > 0.75, OR > 1.04, and P value < 0.01 in all cohorts). In conclusion, IRS is significantly associated with melanoma metastasis and can be a novel effective signature for predicting the metastasis risk.\n",
      "Ping Liang, Yi Chai, He Zhao, Guihuai Wang\n",
      "\n",
      "Glioblastoma (GBM), the most common and aggressive brain tumor, has a very poor outcome and high tumor recurrence rate. The immune system has positive interactions with the central nervous system. Despite many studies investigating immune prognostic factors, there is no effective model to identify predictive biomarkers for GBM. Genomic data and clinical characteristic information of patients with GBM were evaluated by Kaplan-Meier analysis and proportional hazard modeling. Deseq2 software was used for differential expression analysis. Immune-related genes from ImmPort Shared Data and the Cistrome Project were evaluated. The model performance was determined based on the area under the receiver operating characteristic (ROC) curve. CIBERSORT was used to assess the infiltration of immune cells. The results of differential expression analyses showed a significant difference in the expression levels of 2942 genes, comprising 1338 upregulated genes and 1604 downregulated genes (p < 0.05). A population of 24 immune-related genes that predicted GBM patient survival was identified. A risk score model established on the basis of the expressions of the 24 immune-related genes was used to evaluate a favorable outcome of GBM. Further validation using the ROC curve confirmed the model was an independent predictor of GBM (AUC = 0.869). In the GBM microenvironment, eosinophils, macrophages, activated NK cells, and follicular helper T cells were associated with prognostic risk. Our study confirmed the importance of immune-related genes and immune infiltrates in predicting GBM patient prognosis.\n",
      "Xiaoshun Shi, Ruidong Li, Xiaoying Dong, Allen Menglin Chen, Xiguang Liu, Di Lu, Siyang Feng, He Wang, Kaican Cai\n",
      "\n",
      "Tumour cells interfere with normal immune functions by affecting the expression of some immune-related genes, which play roles in the prognosis of cancer patients. In recent years, immunotherapy for tumours has been widely studied, but a practical prognostic model based on immune-related genes in lung adenocarcinoma comparable to existing model has not been established and reported.\n",
      "Minghui Zhang, Kaibin Zhu, Haihong Pu, Zhuozhong Wang, Hongli Zhao, Jinfeng Zhang, Yan Wang\n",
      "\n",
      "We investigated the local immune status and its prognostic value in lung adenocarcinoma. In total, 513 lung adenocarcinoma samples from TCGA and ImmPort databases were collected and analyzed. The R package coxph was employed to mine immune-related genes that were significant prognostic indicators using both univariate and multivariate analyses. The R software package glmnet was then used for Lasso Cox regression analysis, and a prognosis prediction model was constructed for lung adenocarcinoma; clusterProfiler was selected for functional gene annotations and KEGG enrichment analysis. Finally, correlations between the RiskScore and clinical features or signaling pathways were established. Sixty-four immune-related genes remarkably correlated with patient prognosis and were further applied. Samples were hierarchically clustered into two subgroups. Accordingly, the LASSO regression algorithm was employed to screen the 14 most representative immune-related genes (PSMD11, PPIA, MIF, BMP5, DKK1, PDGFB, ANGPTL4, IL1R2, THRB, LTBR, TNFRSF1, TNFRSF17, IL20RB, and MC1R) with respect to patient prognosis. Then, the prognosis prediction model for lung adenocarcinoma patients (namely, the RiskScore equation) was constructed, and the training set samples were incorporated to evaluate the efficiency of this model to predict and classify patient prognosis. Subsequently, based on functional annotations and KEGG pathway analysis, the 14 immune-related genes were mainly enriched in pathways closely associated with lung adenocarcinoma and its immune microenvironment, such as cytokine-cytokine receptor interaction and human T-cell leukemia virus 1 infection. Furthermore, correlations between the RiskScore and clinical features of the training set samples and signaling pathways (such as p53, cell cycle, and DNA repair) were also demonstrated. Finally, the test set sample data were employed for independent testing and verifying the model. We established a prognostic prediction RiskScore model based on the expression profiles of 14 immune-related genes, which shows high prediction accuracy and stability in identifying immune features. This could provide clinical guidance for the diagnosis and prognosis of different immunophenotypes, and suggest multiple targets for precise advanced lung adenocarcinoma therapy based on subtype-specific immune molecules.\n",
      "Bangbei Wan, Bo Liu, Yuan Huang, Gang Yu, Cai Lv\n",
      "\n",
      "Clear cell renal cell carcinoma (ccRCC) is the most common pathological subtype of renal cell carcinoma, and immune-related genes (IRGs) are key contributors to its development. In this study, the gene expression profiles and clinical data of ccRCC patients were downloaded from The Cancer Genome Atlas database and the cBioPortal database, respectively. IRGs were obtained from the ImmPort database. We analyzed the expression of IRGs in ccRCC, and discovered 681 that were differentially expressed between ccRCC and normal kidney tissues. Univariate Cox regression analysis was used to identify prognostic differentially expressed IRGs (PDEIRGs). Using Lasso regression and multivariate Cox regression analyses, we detected seven optimal PDEIRGs (PLAU, ISG15, IRF9, ARG2, RNASE2, SEMA3G and UCN) and used them to construct a risk model to predict the prognosis of ccRCC patients. This model accurately stratified patients with different survival outcomes and precisely identified patients with different mutation burdens. Our findings suggest the seven PDEIRGs identified in this study are valuable prognostic predictors in ccRCC patients. These genes could be used to investigate the developmental mechanisms of ccRCC and to design individualized treatments for ccRCC patients.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Qian Song, Jun Shang, Zuyi Yang, Lanlin Zhang, Chufan Zhang, Jianing Chen, Xianghua Wu\n",
      "\n",
      "Lung cancer has become the most common cancer type and caused the most cancer deaths. Lung adenocarcinoma (LUAD) is one of the major type of lung cancer. This study aimed to establish a signature based on immune related genes that can predict patients' OS for LUAD.\n",
      "Sipeng Shen, Guanrong Wang, Ruyang Zhang, Yang Zhao, Hao Yu, Yongyue Wei, Feng Chen\n",
      "\n",
      "Ovarian cancer (OV) is the most lethal gynecological cancer in women. We aim to develop a generalized, individualized immune prognostic signature that can stratify and predict overall survival for ovarian cancer.\n",
      "Zhongyu Wang, Qian Song, Zuyi Yang, Jianing Chen, Jun Shang, Wen Ju\n",
      "\n",
      "The kidney renal papillary cell carcinoma (KIRP) is a relatively rare type of kidney cancer. There has been no investigation to find a robust signature to predict the survival outcome of KIRP patients in the aspect of tumor immunology. In this study, 285 KIRP samples from The Cancer Genome Atlas (TCGA) were randomly divided into training and testing set. A total of 1534 immune-related genes from The Immunology Database and Analysis Portal (ImmPort) were used as candidates to construct the signature. Using univariate Cox analysis, we evaluated the relationship between overall survival and immune-related genes expression and found 272 immune-related genes with predicting prognostic ability. In order to construct an efficient predictive model, the Cox proportional hazards model with an elastic-net penalty was used and identified 23 groups after 1000 iterations. As a result, 15-genes model showing more stable than other gene groups was chosen to construct our immune-related risk signature. In line with our expectations, the signature can independently predict the survival outcome of KIRP patients. Patients with high-immune risk were found correlated with advanced stage. We also found that the high-immune risk patients with higher PBRM1 and SETD2 mutations, increasing chromosomal instability, together with the gene set enrichment analysis (GSEA) results showing intensive connection of our signature with immune pathways. In conclusion, our study constructs a robust 15-gene signature for predicting KIRP patients' survival outcome on the basis of tumor immune environment and may provide possible relationship between prognosis and immune-related biological function.\n",
      "Marina Sirota, Cristel G Thomas, Rebecca Liu, Maya Zuhl, Payal Banerjee, Ronald J Wong, Cecele C Quaintance, Rita Leite, Jessica Chubiz, Rebecca Anderson, Joanne Chappell, Mara Kim, William Grobman, Ge Zhang, Antonis Rokas, Sarah K England, Samuel Parry, Gary M Shaw, Joe Leigh Simpson, Elizabeth Thomson, Atul J Butte\n",
      "\n",
      "Preterm birth, or the delivery of an infant prior to 37 weeks of gestation, is a significant cause of infant morbidity and mortality. In the last decade, the advent and continued development of molecular profiling technologies has enabled researchers to generate vast amount of 'omics' data, which together with integrative computational approaches, can help refine the current knowledge about disease mechanisms, diagnostics, and therapeutics. Here we describe the March of Dimes' Database for Preterm Birth Research (http://www.immport.org/resources/mod), a unique resource that contains a variety of 'omics' datasets related to preterm birth. The database is open publicly, and as of January 2018, links 13 molecular studies with data across tens of thousands of patients from 6 measurement modalities. The data in the repository are highly diverse and include genomic, transcriptomic, immunological, and microbiome data. Relevant datasets are augmented with additional molecular characterizations of almost 25,000 biological samples from public databases. We believe our data-sharing efforts will lead to enhanced research collaborations and coordination accelerating the overall pace of discovery in preterm birth research.\n",
      "Alexander D Diehl, Terrence F Meehan, Yvonne M Bradford, Matthew H Brush, Wasila M Dahdul, David S Dougall, Yongqun He, David Osumi-Sutherland, Alan Ruttenberg, Sirarat Sarntivijai, Ceri E Van Slyke, Nicole A Vasilevsky, Melissa A Haendel, Judith A Blake, Christopher J Mungall\n",
      "\n",
      "The Cell Ontology (CL) is an OBO Foundry candidate ontology covering the domain of canonical, natural biological cell types. Since its inception in 2005, the CL has undergone multiple rounds of revision and expansion, most notably in its representation of hematopoietic cells. For in vivo cells, the CL focuses on vertebrates but provides general classes that can be used for other metazoans, which can be subtyped in species-specific ontologies.\n",
      "Krithika Bhuvaneshwar, Anas Belouali, Varun Singh, Robert M Johnson, Lei Song, Adil Alaoui, Michael A Harris, Robert Clarke, Louis M Weiner, Yuriy Gusev, Subha Madhavan\n",
      "\n",
      "G-DOC Plus is a data integration and bioinformatics platform that uses cloud computing and other advanced computational tools to handle a variety of biomedical BIG DATA including gene expression arrays, NGS and medical images so that they can be analyzed in the full context of other omics and clinical information.\n",
      "Yu Qian, Chungwen Wei, F Eun-Hyung Lee, John Campbell, Jessica Halliley, Jamie A Lee, Jennifer Cai, Y Megan Kong, Eva Sadat, Elizabeth Thomson, Patrick Dunn, Adam C Seegmiller, Nitin J Karandikar, Christopher M Tipton, Tim Mosmann, Iñaki Sanz, Richard H Scheuermann\n",
      "\n",
      "Advances in multiparameter flow cytometry (FCM) now allow for the independent detection of larger numbers of fluorochromes on individual cells, generating data with increasingly higher dimensionality. The increased complexity of these data has made it difficult to identify cell populations from high-dimensional FCM data using traditional manual gating strategies based on single-color or two-color displays.\n",
      "Mark A Musen, Carol A Bean, Kei-Hoi Cheung, Michel Dumontier, Kim A Durante, Olivier Gevaert, Alejandra Gonzalez-Beltran, Purvesh Khatri, Steven H Kleinstein, Martin J O'Connor, Yannick Pouliot, Philippe Rocca-Serra, Susanna-Assunta Sansone, Jeffrey A Wiser\n",
      "\n",
      "The Center for Expanded Data Annotation and Retrieval is studying the creation of comprehensive and expressive metadata for biomedical datasets to facilitate data discovery, data interpretation, and data reuse. We take advantage of emerging community-based standard templates for describing different kinds of biomedical datasets, and we investigate the use of computational techniques to help investigators to assemble templates and to fill in their values. We are creating a repository of metadata from which we plan to identify metadata patterns that will drive predictive data entry when filling in metadata templates. The metadata repository not only will capture annotations specified when experimental datasets are initially created, but also will incorporate links to the published literature, including secondary analyses and possible refinements or retractions of experimental interpretations. By working initially with the Human Immunology Project Consortium and the developers of the ImmPort data repository, we are developing and evaluating an end-to-end solution to the problems of metadata authoring and management that will generalize to other data-management environments. \n",
      "Y Megan Kong, Carl Dahlke, Qun Xiang, Yu Qian, David Karp, Richard H Scheuermann\n",
      "\n",
      "Clinical research includes a wide range of study designs from focused observational studies to complex interventional studies with multiple study arms, treatment and assessment events, and specimen procurement procedures. Participant characteristics from case report forms need to be integrated with molecular characteristics from mechanistic experiments on procured specimens. In order to capture and manage this diverse array of data, we have developed the Ontology-Based eXtensible data model (OBX) to serve as a framework for clinical research data in the Immunology Database and Analysis Portal (ImmPort). By designing OBX around the logical structure of the Basic Formal Ontology (BFO) and the Ontology for Biomedical Investigations (OBI), we have found that a relatively simple conceptual model can represent the relatively complex domain of clinical research. In addition, the common framework provided by BFO makes it straightforward to develop data dictionaries based on reference and application ontologies from the OBO Foundry.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Randi Vita, James A Overton, Patrick Dunn, Kei-Hoi Cheung, Steven H Kleinstein, Alessandro Sette, Bjoern Peters\n",
      "\n",
      "An Immune Exposure is the process by which components of the immune system first encounter a potential trigger. The ability to describe consistently the details of the Immune Exposure process was needed for data resources responsible for housing scientific data related to the immune response. This need was met through the development of a structured model for Immune Exposures. This model was created during curation of the immunology literature, resulting in a robust model capable of meeting the requirements of such data. We present this model with the hope that overlapping projects will adopt and or contribute to this work.\n",
      "Harry Pickering, Andy Teng, Nkoyo Faal, Hassan Joof, Pateh Makalo, Eunice Cassama, Meno Nabicassa, Anna R Last, Sarah E Burr, Sarah L Rowland-Jones, Nicholas R Thomson, Chrissy H Roberts, David C W Mabey, Robin L Bailey, Richard D Hayward, Luis M de la Maza, Martin J Holland\n",
      "\n",
      "The frequency and duration of Chlamydia trachomatis (Ct) ocular infections decrease with age, suggesting development of partial immunity. However, there is a lack of clear correlates of immunity to Ct infection in humans. We screened sera from a cohort of Gambian children followed for six-months against a Ct-proteome microarray. At genome sequence level, we detected signatures of selection from a population of ocular Ct isolates from Guinea-Bissau. Together these approaches allowed us to highlight the focus of humoral responses and hypothesise new modes of pathogen immune evasion. Children who were susceptible to frequent and/or prolonged Ct infection had a less focussed antibody response, including preferential recognition of forty-two antigens. There was evidence of positive and purifying selection across the genome, but little balancing selection. In contrast, most antigens that were associated with susceptibility were under neutral selection. These data suggest an evasion strategy in which Ct presents a large panel of irrelevant antigens to the immune system to block or misdirect protective responses. Development of a focused immune response, possibly induced through vaccination, may be an effective strategy to promote protection to Ct infection.\n",
      "Kaustuv Nayak, Lichen Jing, Ronnie M Russell, D Huw Davies, Gary Hermanson, Douglas M Molina, Xiaowu Liang, David R Sherman, William W Kwok, Junbao Yang, John Kenneth, Syed F Ahamed, Anmol Chandele, Kaja Murali-Krishna, David M Koelle\n",
      "\n",
      "Elicitation of CD4 IFN-gamma T cell responses to Mycobacterium tuberculosis (MTB) is a rational vaccine strategy to prevent clinical tuberculosis. Diagnosis of MTB infection is based on T-cell immune memory to MTB antigens. The MTB proteome contains over four thousand open reading frames (ORFs). We conducted a pilot antigen identification study using 164 MTB proteins and MTB-specific T-cells expanded in vitro from 12 persons with latent MTB infection. Enrichment of MTB-reactive T-cells from PBMC used cell sorting or an alternate system compatible with limited resources. MTB proteins were used as single antigens or combinatorial matrices in proliferation and cytokine secretion readouts. Overall, our study found that 44 MTB proteins were antigenic, including 27 not previously characterized as CD4 T-cell antigens. Antigen truncation, peptide, NTM homology, and HLA class II tetramer studies confirmed malate synthase G (encoded by gene Rv1837) as a CD4 T-cell antigen. This simple, scalable system has potential utility for the identification of candidate MTB vaccine and biomarker antigens.\n",
      "Dorothy L Patton, Andy Teng, Arlo Randall, Xiaowu Liang, Philip L Felgner, Luis M de la Maza\n",
      "\n",
      "The cervix and/or fallopian tubes of pigtailed macaques were experimentally infected with Chlamydia trachomatis. Their sera were collected at varying time points and screened for identification of immunodominant antigens using a whole-genome protein microarray. The effect of doxycycline treatment on the antibody response generated in these macaques was also investigated. Twenty-five female macaques were infected with C. trachomatis serovars D or E in the cervix and/or fallopian tubes. Bloods were collected at baseline and at various intervals after challenge. Serum samples were tested for antibodies using a C. trachomatis serovar D protein microarray. Twenty chlamydial antigens reacted with sera from at least 68% (17/25) of the macaques. In addition to some well-known chlamydial antigens, nine different proteins, not previously recognized as immunodominant, including four hypothetical proteins (CT005, CT066, CT360 and CT578), were identified. Monkeys infected in the fallopian tubes developed a more robust antibody response than animals inoculated in the cervix. Treatment with doxycycline significantly decreased Chlamydia-specific antibody levels. In summary, using protein microarray serum samples from experimentally infected pigtailed macaques were screened for immunodominant chlamydial antigens. These antigens can now be tested in animal models for their ability to protect and as markers of disease progression.\n",
      "Chunmei Cheng, Pooja Jain, Sukumar Pal, Delia Tifrea, Guifeng Sun, Andy A Teng, Xiaowu Liang, Philip L Felgner, Luis M de la Maza\n",
      "\n",
      "A novel Chlamydia muridarum antigen (TC0582) was used to vaccinate BALB/c mice. Mice were also immunized with other components of the ATP synthase complex (TC0580, TC0581, and TC0584), or with the major outer membrane protein (MOMP). TC0582 was also formulated in combination with TC0580, TC0581 or MOMP. TC0582 alone, or in combination with the other antigens, elicited strong Chlamydia-specific humoral and cellular immune responses. Vaccinated animals were challenged intranasally and the course of the infection was followed for 10 days. Based on percentage change in body weight, lung weight, and number of Chlamydia inclusion forming units recovered from the lungs, mice immunized with TC0582, TC0581 or MOMP, as single antigens, showed significant protection. Mice immunized with combinations of two antigens were also protected but the level of protection was not additive. TC0582 has sequence homology with the eukaryotic ATP synthase subunit A (AtpA). Therefore, to determine if immunization with TC0582, or with Chlamydia, elicited antibodies that cross-reacted with the mouse AtpA, the two proteins were printed on a microarray. Sera from mice immunized with TC0582 and/or live Chlamydia, strongly reacted with TC0582 but did not recognize the mouse AtpA. In conclusion, TC0582 may be considered as a Chlamydia vaccine candidate. \n",
      "Andy Teng, Maria I Cruz-Fisher, Chunmei Cheng, Sukumar Pal, Guifeng Sun, Pooja Ralli-Jain, Douglas M Molina, Philip L Felgner, Xiaowu Liang, Luis M de la Maza\n",
      "\n",
      "Chlamydia trachomatis is the most common bacterial sexually transmitted pathogen in the world. To identify new vaccine candidates a protein microarray was constructed by expressing the open reading frames (ORFs) from Chlamydia mouse pneumonitis (MoPn). C57BL/6, C3H/HeN and BALB/c mice were immunized either intranasally or intravaginally with live MoPn elementary bodies (EB). Two additional groups were immunized by the intramuscular plus subcutaneous routes with UV-treated EB, using CpG and Montanide as adjuvants to favor a Th1 response, or Alum, to elicit a Th2 response. Serum samples collected from the three strains of mice were tested in the microarray. The array included the expression of 909 proteins from the 921 ORFs of the MoPn genome and plasmid. A total of 530 ORFs were recognized by at least one serum sample. Of these, 36 reacted with sera from the three strains of mice immunized with live EB. These antigens included proteins that were previously described as immunogenic such as MOMP and HSP60. In addition, we uncovered new immunogens, including 11 hypothetical proteins. In summary, we have identified new immunodominant chlamydial proteins that can be tested for their ability to induce protection in animal models and subsequently in humans.\n",
      "Douglas M Molina, Sukumar Pal, Mathew A Kayala, Andy Teng, Paul J Kim, Pierre Baldi, Philip L Felgner, Xiaowu Liang, Luis M de la Maza\n",
      "\n",
      "Chlamydia trachomatis is the most common bacterial sexually transmitted pathogen in the world. In order to control this infection there is an urgent need to formulate a vaccine. Identification of protective antigens is required to implement a subunit vaccine. To identify potential antigen vaccine candidates, three strains of mice, BALB/c, C3H/HeN and C57BL/6, were inoculated with live and inactivated C. trachomatis mouse pneumonitis (MoPn) by different routes of immunization. Using a protein microarray, serum samples collected after immunization were tested for the presence of antibodies against specific chlamydial antigens. A total of 225 open reading frames (ORF) of the C. trachomatis genome were cloned, expressed, and printed in the microarray. Using this protein microarray, a total of seven C. trachomatis dominant antigens were identified (TC0052, TC0189, TC0582, TC0660, TC0726, TC0816 and, TC0828) as recognized by IgG antibodies from all three strains of animals after immunization. In addition, the microarray was probed to determine if the antibody response exhibited a Th1 or Th2 bias. Animals immunized with live organisms mounted a predominant Th1 response against most of the chlamydial antigens while mice immunized with inactivated Chlamydia mounted a Th2-biased response. In conclusion, using a high throughput protein microarray we have identified a set of novel proteins that can be tested for their ability to protect against a chlamydial infection.\n"
     ]
    }
   ],
   "source": [
    "all_articles = []\n",
    "for articleID in article_ids:\n",
    "   \n",
    "    get_url = \"http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=pubmed&retmode=xml&id=\"\n",
    "    get_url = get_url + articleID\n",
    "    \n",
    "    _create_unverified_https_context = ssl._create_unverified_context\n",
    "    response = requests.get(get_url)\n",
    "\n",
    "    soup_object = BeautifulSoup(response.content, \"xml\")\n",
    "    one_article = get_elements(soup_object)\n",
    "    all_articles.append(one_article)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Converts the gathered data and exports it as a CSV file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_frame = pd.DataFrame(all_articles)\n",
    "data_frame.columns = [\"Authors\", \"ArticleTitle\", \"JournalTitle\", \"JournalVolume\", \"JournalDate\", \"PUBID\", \"Abstact\"]\n",
    "\n",
    "csv_name = keyword + \".csv\"\n",
    "data_frame.to_csv(csv_name)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
